pil-analyzer/src/pil_analyzer.rs

```diff
@@ -257,6 +257,113 @@ impl PILAnalyzer {
         }
     }
 
+    /// Verifies that all struct instantiations match their corresponding declarations (existence of field names, completeness)
+    fn struct_fields_check(&self) -> Result<(), Vec<Error>> {
+        let mut errors = Vec::new();
+        let mut visited_structs = HashSet::new();
+
+        let structs_exprs = self.all_children().filter_map(|expr| {
+            if let Expression::StructExpression(
+                sr,
+                StructExpression {
+                    name: Reference::Poly(poly),
+                    fields,
+                },
+            ) = expr
+            {
+                Some((sr, poly.name.clone(), fields))
+            } else {
+                None
+            }
+        });
+
+        for (sr, name, fields) in structs_exprs {
+            if let Some((
+                _,
+                Some(FunctionValueDefinition::TypeDeclaration(TypeDeclaration::Struct(
+                    struct_decl,
+                ))),
+            )) = self.definitions.get(&name)
+            {
+                let declared_fields: HashSet<_> =
+                    struct_decl.fields.iter().map(|f| &f.name).collect();
+                let used_fields: HashSet<_> = fields.iter().map(|f| &f.name).collect();
+
+                errors.extend(
+                    fields
+                        .iter()
+                        .filter(|f| !declared_fields.contains(&f.name))
+                        .map(|f| {
+                            sr.with_error(format!(
+                                "Struct '{name}' has no field named '{}'",
+                                f.name
+                            ))
+                        }),
+                );
+
+                errors.extend(declared_fields.difference(&used_fields).map(|&f| {
+                    sr.with_error(format!("Missing field '{f}' in initializer of '{name}'",))
+                }));
+
+                let duplicate_fields: Vec<_> =
+                    fields.iter().map(|f| &f.name).duplicates().collect();
+
+                errors.extend(
+                    duplicate_fields
+                        .into_iter()
+                        .map(|f| sr.with_error(format!("Field '{f}' specified more than once"))),
+                );
+            } else {
+                errors.push(sr.with_error(format!("Struct '{name}' has not been declared")));
```

Comment:
Can this happen? Shouldn't it be a panic? Or should it rather be an error that we exected a struct and got something else?